src/lib.rs

```diff
@@ -1,91 +1,64 @@
-use cfgrammar::RIdx;
-use lrlex::{lrlex_mod, DefaultLexeme, DefaultLexerTypes, LRNonStreamingLexerDef};
-use lrpar::{lrpar_mod, Lexeme, Node};
+use lrlex::{lrlex_mod, DefaultLexerTypes};
+use lrpar::{lrpar_mod, NonStreamingLexer};
 
+// Using `lrlex_mod!` brings the lexer for `calc.l` into scope. By default the module name will be
+// `calc_l` (i.e. the file name, minus any extensions, with a suffix of `_l`).
 lrlex_mod!("calc.l");
+// Using `lrpar_mod!` brings the parser for `calc.y` into scope. By default the module name will be
+// `calc_y` (i.e. the file name, minus any extensions, with a suffix of `_y`).
 lrpar_mod!("calc.y");
 
-pub fn eval(
-    lexerdef: &LRNonStreamingLexerDef<DefaultLexerTypes>,
-    input: &str,
-) -> Result<String, String> {
-    let lexer = lexerdef.lexer(input);
-    let (res, errs) = calc_y::parse(&lexer);
-    for e in errs {
-        println!("{}", e.pp(&lexer, &calc_y::token_epp));
-    }
-    match res {
-        Some(x) => {
-            return Ok(format!("{}", Eval::new(input).eval(&x)));
+use calc_y::Expr;
+
+#[derive(Debug)]
+pub struct Calc {}
+
+impl Calc {
+    pub fn from_str(input: &str) -> Result<u64, &'static str> {
```

Comment:
I expected `from_str()` to make your AST and a separate `eval` (or `exec`) method to execute it. Do you think there's any value in separating the two tasks?

Reply:
I tried it in one of my iterations and had some issues with rust types so decided to merge it into one. 
I'll try again :muscle:

Reply:
done. Ast and eval func are separated https://github.com/softdevteam/pavel.yaiwr/pull/8/commits/54ac1dab51d2e4a9aa6fccb0ee70e0abbe336a4f